Converged menu spec (#16517)
microsoft/fluentui: specs/Menu.md

- overflowBoundary - the bounds to shift the popup without overflowing
- mountNode - where the popup is actually rendered in the DOM, by default this is a portal to a div in body

```typescript
<ContextualMenu
  // pixel values for bounding rectangle
  // defaults to target window as default bounding rectangle
  bounds={{height: 0, width: 0, top: 0, left:0 , right: 0, bottom: 0}}
  // callback for bounds
  bounds{(target, targetWindow) => ({/*Same object as above*/})}
  target={htmlElement}

  // renders to a portal node on body
  layerProps={/*ILayerProps*/}

  // every single one of the above can all be declared here too
  calloutProps={{bounds, target}}
/>
```

`ContextualMenu` or `Callout` has no notion of separate boundaries for flip or overflow, and auto behaviour is used for flip and overflow 'pushing'
It should also be noted that `ContextualMenu` allows all of the same props to be passed to its submenus to custom tweak position for each submenu if necessary

#### Submenu positioning

The default positioning for both v0 and v7 submenus is:

- rightTopEdge (v7)
- top-after (v0)

Both will also flip appropriately when the overflow boundary is too small.

> The main difference between the two is that v0 submenu's position does not expose any way to customize or override the positioning of the submenu. However v7 allows every single customization as the root menu. It is very possible to do the below:

Reviewer: I **think** it is possible to pass popper props to submenu. @layershifter is it correct?

Author: Looking at our code I don't think so:

https://github.com/microsoft/fluentui/blob/89eb6bc42cdc8d49277007e2268cd52f0beced6c/packages/fluentui/react-northstar/src/components/Menu/MenuItem.tsx#L504

Author: It is possible but is actually not present in our documentation, it's a hidden prop you can add `menu.popper` https://fluentsite.z22.web.core.windows.net/0.52.0/components/menu/props

Reviewer 2: Just FYI, hidden props mean we failed. Let's ensure we don't do this moving forward.

# Menu

## Background

### Definition

> This spec defines the default function of a `Menu` as an interactive component that displays a list of options that can be represented by a range possible states. Possible variants are defined in [the relevant section](##variants)

Reviewer 3: "the relevant section" link is broken

Author: I think it's fixed :)

The `Menu` should be displayed on a temporary surface that interrupts the normal flow of content. The temporary surface should be triggered by an external user action such as (but not limited to) a click on a button or other UI control.

The interactions that result in the dismiss/removal of the `Menu` component should be configurable.

## Prior art

> As a part of the spec definitions in Fluent UI, a research effort has been made through [Open UI](https://open-ui.org/). The current research proposal is available as an open source contribution undergoing review ([LINK to research proposal](https://github.com/WICG/open-ui/pull/249))

Reviewer 3: ```suggestion
As a part of the spec definitions in Fluent UI, a research effort has been made through [Open UI](https://open-ui.org/). The current research proposal is available as an open source contribution undergoing review ([research proposal](https://github.com/WICG/open-ui/pull/249))
```

nit

## Comparison of `@fluentui/react` and `@fluentui/react-northstar`

> All mentions of v7 or v8 == `@fluentui/react`
> All mentions of v0 == `@fluentui/react-northstar`

Reviewer 3: ![image](https://user-images.githubusercontent.com/14183168/105016624-25e4ea80-5a43-11eb-9dd4-157a120ee1e3.png)

```suggestion
- All mentions of v7 or v8 == `@fluentui/react`
- All mentions of v0 == `@fluentui/react-northstar`
```

formatting?

The most relevant comparison that can be achieved between the two libraries is between `ContextualMenu` in v7 and a combination of `Menu`, `Popup` and `ToolbarItem` in v0.

> v0 suffers from a consistency issue that the control used in `Menu` and the menu variant of `ToolbarItem` are not actually the same component and have different behaviour. However, semantically for the purposes of this spec, they represet the same control that will be implemented.

Reviewer 3: ```suggestion
v0 suffers from a consistency issue that the control used in `Menu` and the menu variant of `ToolbarItem` are not actually the same component and have different behavior. However, semantically for the purposes of this spec, they represent the same control that will be implemented.
```

typos 🤹

Author: I would argue to the death that `behaviour` is the correct spelling, but sure I can change :-)

> Note that the below code samples are not meant to be complete, but to highlight differences between the two libraries. Please refer to official docsites for actual API references.

Reviewer 3: ```suggestion
_Note that the below code samples are not meant to be complete, but to highlight differences between the two libraries. Please refer to official docsites for actual API references._
```

Should we add links there?

Author: Definitely

### Positioning

> `ContextualMenu` in v7 is a component that also exposes the API to control the positioning of the temporary popup surface that the Menu is rendered on. This aspect of the v7 component should be compared with the `Popup` component in v0 since the v0 `Menu` is created as a standalone component with no positioning properties.

Reviewer 3: > since the v0 `Menu` is created as a standalone component with no positioning properties.

We have positioning properties for submenus...

Author: Aaaaah that makes my life complicated. I will resolve this for now and track this issue in the checklist in PR description

<Popup
  trigger={<Button />}
  content={...}
/>
// v0 - children trigger
<Popup content={...}>
  <Button icon={<MoreIcon />} title="Show popup" />
</Popup>
```

### Layout variations

The v7 `ContextualMenu` only has one primary layout which is a vertical list of menu items.

The v0 `Menu` component differs clearly in this that the default layout is a horizontal menu. To achieve the same layout as `ContextualMenu` (and the layout defined in this spec) it's necessary to use the `vertical` prop which is `false` by default.

```typescript
<Menu items={items} vertical />
```

### Open/Dismiss events

The v7 `ContextualMenu` is intended to be used as a controlled component. The visibiltiy of the menu is controlled using the `hidden` prop whose value should be React state of the cosumer. A separate `onDismiss` prop can also be used that will be invoked during events where the callout tries to close, i.e. click outside the content.

The v0 `Popup` should be compared here, since the v0 `Menu` does not handle open/dismiss events. `Popup` visibility can be controlled using the `open` prop. `Popup` provides a callback prop `onOpenChange` which can be used both to open and dismiss.

> However as mentioned above, the general usage of `Popup` is with an autocontrolled trigger. If a trigger is used with the `open` prop, this disables autocontrolling

Reviewer 3: I suggest to rephrase this and say that `Popup` implements autoconrolled pattern that allows to use components as controlled and uncontrolled: https://reactjs.org/docs/uncontrolled-components.html

Author: Reworded and fixed the labelling of the examples

  <Menu.Item index={0}>
    <Menu.ItemContent>Editorials</Menu.ItemContent>
  </Menu.Item>
</Menu>

// v0 autocontrolled active index controlled
<Menu activeIndex={0}>
  <Menu.Item index={0}>
    <Menu.ItemContent>Editorials</Menu.ItemContent>
  </Menu.Item>
</Menu>

// shorthand variation
const items = [
  {
    key: 'editorials',
    content: 'Editorials',
  },
]
<Menu defaultActiveIndex={0} items={items} />
```

In order to obtain semantically meaningful selection state in v0, the only possible way is to use a `Toolbar` component. The menu that is rendered in this component is completely different but supports both checkbox and radio selection states through the use of an `active` prop and must be controlled.

```typescript
// Toolbar with one item that opens a selectable menu
const toolbarItems = [
  {
    icon: <MoreIcon />,
    title: 'More',
    menu: [
      {
        active: true,
        content: 'Bold',
        kind: 'toggle',
        // kind: 'radio', // for radio
        index: 0,
        onClick: handleToggleClick,
      },
      {
        active: false,
        content: 'Italic',
        kind: 'toggle',
        // kind: 'radio', // for radio

>         index: 1,

Reviewer 3: `index` is not required for shorthands

Author: removed for shorthands

    {onRenderIcon: (props: IContextualMenuItemProps, defaultRenderer) => {}},
  ]
}

<ContextualMenu menuProps={menuProps}>
```

Custom data can also be associated with menu items

```typescript
const menuProps = {
  items: [{
    ...
    data: { foo: "bar" }
  }]
}
```

v0 custom rendering through shorthand components is a consistent experience through all shorthand components, but provide a smaller API surface (whether this is simpler or less powerful can be subjective). Custom rendering in the case of the `Menu` component would be done through the use of `chilren` prop either through the standard React child component API or through shorthand as a callback function.

```typescript
// v0 shorthand children render callback
const items = [
  {
    key: 'editorials',
    children: (El, props) => <El>{props.key}</El>
  },
]

<Menu defaultActiveIndex={0} items={items} />

// v0 children API custom render
<Menu>
  <Menu.Item index={0}>
    <Menu.ItemContent>Editorials</Menu.ItemContent>
  </Menu.Item>
  <Menu.Item index={1}>
    CustomContent
  </Menu.Item>
  {/*Not recommended but definitely possible*/}
  <div>custom item</div>
</Menu>
```

## Variants

> ### Nested nenus

Reviewer 3: ```suggestion
### Nested menus
```

### MenuSection

Creates a section inside a `MenuList`, setting up header layout and dividers between `MenuItems`.

The MenuSection is also a useful component to declare different selection groups (checkbox/radio) in a `MenuList`.

### MenuItem

As the name infers

### SubMenu

Creates a `Menu` component with `MenuItem` trigger and handles the positioning of the nested menu.

## Sample code

The below samples do not represent the definitive props of the final implemented component, but represent the ideal final implementations. Can be subject to change during the implementation phase.

### Default Menu

```typescript
const trigger = <button> Open menu </button>

const menu = (
  <Menu trigger={trigger}>
    <MenuItem title="Option 1" />
    <MenuItem title="Option 2" />
    <MenuItem title="Option 3" />
  <Menu>
)
```

### Sections and submenus

```typescript
const trigger = <button> Open menu </button>

const menu = (

>   <Menu trigger={trigger}>

Reviewer 3: In the prototype we intentionally skipped `trigger` as it breaks JSX layout:

```jsx
<Menu>
  <MenuItem>
  {/* nested item ⬇️ */}
  <Menu trigger={<MenuItem />> {/* what will be rendered in the layout ❔ */}
    <MenuItem />
  </>
</Menu>
```

Author: My suggestion was to have a separate `SubMenu` component in these situations. The root level `Menu` for me only exists as sugar for positioning

Reviewer 4: Have you considered some kind of "MenuAnchor" wrapper element that wraps the trigger and has a slot for the menu? This is just brainstorming, but perhaps something like:
```tsx
<MenuAnchor menu={
    <Menu>
      <MenuItem>Item 1</MenuItem>
      <MenuItem>Item 2</MenuItem>
    </Menu>}>
  <Button>Open Menu</Button>
</MenuAnchor>
```

That might also be useful for separating the responsibility for creating and positioning the popup (handled by MenuAnchor), and rendering the content of the menu (handled by Menu/MenuItem). That way a menu could be rendered inside another container that's not a popup (or a custom popup).

Author: @behowell  I would avoid wrapping the trigger and putting the menu in a slot. This is because I would like `Menu` to reflect correct dom structure with JSX as much as possible to have an API consistent with all the other menu implementations in react that are out there.

Please feel free to join the trigger deep dive and weigh in with your ideas

    <MenuItem title="Option 1" />
    <MenuSection title="Section">
      <MenuItem title="Section Option 1">
      <MenuItem title="Section Option 2">
      <MenuItem title="Section Option 3">
    <MenuSection />
    <SubMenu title="Open submenu">
      <MenuItem title="Submenu Option 1">
      <MenuItem title="Submenu Option 2">
      <MenuItem title="Submenu Option 3">
    <SubMenu>
  <Menu>
)
```

### Standlone

```typescript
const [open] = React.useState(false);

const menu = (
  <CustomSurface open={open}>
    <MenuList>
      <MenuItem title="Option 1" />
      <MenuItem title="Option 2" />
      <MenuItem title="Option 3" />
    <MenuList>
  <CustomSurface>
)
```

### Selection

```typescript
const trigger = <button> Open menu </button>
const [selectedItems, setSelectedItems] = React.useState([]);

// basic checkbox example
const menuCheckbox = (
  <Menu
    selectedItems={selectedItems}
    onSelectionChange={setSeelctedItems}
    trigger={trigger}
  >
    <MenuItem kind="checkbox" key={1} title="Option 1">
    <MenuItem kind="checkbox" key={2} title="Option 2">

>     <MenuItem kind="checkbox" key={3} title="Option 3">

Reviewer 3: `index` in the prototype is used intentionally as `key` is not a React prop and can't be accessed inside a component

Author: Changed for `index` might play around with the prototype and have a think. But as I keep stating, this spec does not set the implementation in stone

/>
```

The v0 `Popup` component has an alternative prop, `trigger`, which accepts a React component. This prop simplifies the creation of temporary content by autocontrolling the open/dismiss functionality.

```typescript
// v0 - shorthand trigger
<Popup
  trigger={<Button />}
  content={...}
/>
// v0 - children trigger
<Popup content={...}>
  <Button icon={<MoreIcon />} title="Show popup" />
</Popup>
```

### Layout variations

The v7 `ContextualMenu` only has one primary layout which is a vertical list of menu items.

The v0 `Menu` component differs clearly in this that the default layout is a horizontal menu. To achieve the same layout as `ContextualMenu` (and the layout defined in this spec) it's necessary to use the `vertical` prop which is `false` by default.

```typescript
<Menu items={items} vertical />
```

### Open/Dismiss events

The v7 `ContextualMenu` is intended to be used as a controlled component. The visibiltiy of the menu is controlled using the `hidden` prop whose value should be React state of the cosumer. A separate `onDismiss` prop can also be used that will be invoked during events where the callout tries to close, i.e. click outside the content.

> The v0 `Popup` should be compared here, since the v0 `Menu` does not handle open/dismiss events. `Popup` visibility can be controlled using the `open` prop. `Popup` provides a callback prop `onOpenChange` which can be used both to open and dismiss.

Reviewer: I think there is open/dismiss for submenus in v0

Author: Yup, I overlooked that, will be looking to incorporate this

As mentioned above, `Popup` implements an autocontrolled pattern which allows both controlled an uncontrolled variants to be used in its API.

```typescript
// v7 controlled ContextualMenu
const [showContextualMenu, setShowContextualMenu] = React.useState(false);
const onShowContextualMenu = () => setShowContextualMenu(true);
const onHideContextualMenu = () => setShowContextualMenu(false);

<ContextualMenu
  hidden={!showContextualMenu}
  onItemClick={onHideContextualMenu}
  onDismiss={onHideContextualMenu
/>;

// v0 uncontrolled Popup
const [open, setOpen] = React.useState(false);

<Popup
  onOpenChange={(e, props: PopupProps) => {/*react on changes*/}}
  trigger={<Button icon={<OpenOutsideIcon />} title="Open popup" />}
/>;

// v0 controlled Popup - used with trigger disables autocontrol
const [open, setOpen] = React.useState(false);

<Popup
  open={open}
  onOpenChange={(e, props: PopupProps) => setOpen(!props.open)}
  trigger={<Button icon={<OpenOutsideIcon />} title="Open popup" />}
/>;
```

### Keyboard navigation

> Both v7 and v0 support arrow navigation in the menu, and home/end keys to jump to first and last items respectively.

Reviewer: there is also quite complex interaction with regards to left/right arrow keys related to opening and closing submenus described by ARIA which v0 implements.

> One interesting difference is that the v7 `ContextualMenu` will also tab through items. The v0 `Menu` on the other hand uses tab to focus in/out of the entire component.

Reviewer: According to ARIA, Menu should not handle Tab - Tab should move focus to the next element (outside of the menu). We might need to keep Tab handling as optional for backward compatibility with v7

Author: That's an interesting point to keeps this backwards compatibility, I'll it to the behaviour spec

> `ContextualMenu` will also allow disabled items to be focusable while navigating the list, while the v0 `Menu` does not permit this.

Reviewer: we have a requrement and a backlog feature to make disabled menu items focusable by default.

Reviewer 5: @jurokapsiar - Does this apply to menu items only or to any element which is focusable when not disabled?

### Selection state

> The v0 `Menu` component supports and `active` state and has a number of props to manage this state. However, this state only affects items visually and does not perform the same function as menu item checkboxes or radio items. The `active` state of menu items can both be controlled and autocontrolled

Reviewer: The active state is confusing and should be removed I think - given that Tabs and Toolbar will be a separate component.

Reviewer: however there are edge cases where menu item checkbox and menu item radio might have active state

Author: I agree, I was super confused trying to figure out in our own codebase what this `active` state actually was.

Author: I make no mention of this in the proposed behaviour/api so far and do not intend to

      text: 'Share',
      canCheck: true,
      isChecked: false,
      onClick: onToggleSelect,
    },
  ],
};

// shorthand usage in a menu button
<DefaultButton menuProps={menuProps} />;
```

### Custom rendering and data

v7 provides render callbacks that can be used to render either the entire menu list or specific slots of menut items. Each call back provides the props avaialble to that slot and a `defaultRender` which allows to easily extend the original render, if required.

```typescript
// v7 custom rendering
const menuProps = {
  onRenderMenuList: (props: IContextualMenuListProps, defaultRenderer) => {},
  onRenderSubMenu: (props: IContextualMenuProps, defaultRenderer) => {},
  items: [
    {
      onRender: (
        item: any,
        dismissMenu: (ev?: any, dismissAll?: boolean) => void
      ) => React.ReactNode
    }
    {onRenderContent: (props: IContextualMenuItemProps, defaultRenderer) => {}},
    {onRenderIcon: (props: IContextualMenuItemProps, defaultRenderer) => {}},
  ]
}

<ContextualMenu menuProps={menuProps}>
```

Custom data can also be associated with menu items

```typescript
const menuProps = {
  items: [{
    ...
    data: { foo: "bar" }
  }]
}
```

> v0 custom rendering through shorthand components is a consistent experience through all shorthand components, but provide a smaller API surface (whether this is simpler or less powerful can be subjective). Custom rendering in the case of the `Menu` component would be done through the use of `chilren` prop either through the standard React child component API or through shorthand as a callback function.

Reviewer: ```suggestion
v0 custom rendering through shorthand components is a consistent experience through all shorthand components, but provide a smaller API surface (whether this is simpler or less powerful can be subjective). Custom rendering in the case of the `Menu` component would be done through the use of `children` prop either through the standard React child component API or through shorthand as a callback function.
```

]

<Menu defaultActiveIndex={0} items={items} />

// v0 children API custom render
<Menu>
  <Menu.Item index={0}>
    <Menu.ItemContent>Editorials</Menu.ItemContent>
  </Menu.Item>
  <Menu.Item index={1}>
    CustomContent
  </Menu.Item>
  {/*Not recommended but definitely possible*/}
  <div>custom item</div>
</Menu>
```

## Variants

### Nested menus

A `Menu` should be able to trigger an additional instance of itself as a part of one or more of its options. The nested `Menu` component should have the same functional capabilities as the root `Menu` component.

The actions that trigger the the nested `Menu` should be be consistent with the actions that can trigger any root `Menu` from a similar UI control.

We advise that no more than two nested `Menu` components be used, but this spec does not functionally apply that constrain to the implementation of the `Menu` component.

### Selection state

A `Menu` should be able to track and represent the selection state of all or some of its options if required.

When an options is associated with a selection state. The `Menu`, either root or nested, should control its dismiss behavior accordingly based on configuration.

### Sections

> A `Menu` can be partitioned into sections using visible dividers in its list of options. Each section can contain a heading title that announces or briefly describes the options in the particular section

Reviewer: for accessibility, wrapping sections in a `role='group'` might be necessary to give them a common title. But this needs to be verified.

Author: Yeah i would like run a11y attributes with milan and incorporate that here soon, it's already tracked in the checklist in description

### Secondary text

An option of a `Menu` component should be able to declare additional secondary text that can provide additional context describing the option or its usage.

For example a secondary text can be a label that shows a keyboard shortcut that will perform an equivalent action of the option of the `Menu` component.

### Split option with nesting

An option of a `Menu` component can trigger a nested `Menu` component and also perform its default action by splitting the option into two interactable areas that handle each action separately.

### Disabled option(s)

All options in a `Menu` component can be disabled and should provide a visible indication for this purpose. User interaction should be defined for disabled options

### Scrollable

A `Menu` should display a vertical scrollbar when the number of options exceeds the height of the component

### Standalone/No surface

> A `Menu` can be used without the temporary popup surface and its trigger. This will allow `Menu` components to be permanent page content or used in custom surfaces with a wider range of UI components.

Reviewer: I would vote for having this as a separate (sub)component - Menu.List?

Author: There is perhaps too many different layers of abstraction here. In the API section this is exactly why is being proposed. I'm not entirely sure we should map the requirements to actual components

### MenuSection

Creates a section inside a `MenuList`, setting up header layout and dividers between `MenuItems`.

The MenuSection is also a useful component to declare different selection groups (checkbox/radio) in a `MenuList`.

### MenuItem

As the name infers

### SubMenu

Creates a `Menu` component with `MenuItem` trigger and handles the positioning of the nested menu.

## Sample code

The below samples do not represent the definitive props of the final implemented component, but represent the ideal final implementations. Can be subject to change during the implementation phase.

### Default Menu

```typescript
const trigger = <button> Open menu </button>

const menu = (
  <Menu trigger={trigger}>
    <MenuItem title="Option 1" />
    <MenuItem title="Option 2" />
    <MenuItem title="Option 3" />
  <Menu>
)
```

### Sections and submenus

```typescript
const trigger = <button> Open menu </button>

const menu = (
  <Menu trigger={trigger}>

>     <MenuItem title="Option 1" />

Reviewer: `title` collides with HTML `title` attribute. `content` or something similar might be better

Author: Makes sense, work is still needed to define the props a lot better. The sample code usages for now should show the intention of the API design

Author: After more thought, I've converted to using children here.

      <MenuItem index={3} title="Option 3">
    </MenuSection>
    <MenuSection kind="radio">
      <MenuItem index={4} title="Option 1">
      <MenuItem index={5} title="Option 2">
      <MenuItem index={6} title="Option 3">
    </MenuSection>
  <Menu>
)
```

// TODO positioning examples ?

## Behaviors

### Menu open/dismiss

A menu can be triggered by the following user interactions on the triggering/anchor element. Not all interactions should be supported at the same time, but the component must be able to support combinations of the below interactions.

As a general rule, once the menu is closed the focus should return to the triggering element once the menu is closed unless the interaction would involve another focusable element.

| Type     | Action     | Result  | Details                                                                |
| -------- | ---------- | ------- | ---------------------------------------------------------------------- |
| Mouse    | Click      | Open    | Click on the trigger element                                           |
| Mouse    | Hover      | Open    | Hover over the trigger element with delay                              |

> | Mouse    | LongPress  | Open    | MouseDown with delay                                                   |

Reviewer: isn't long press equivalent to right click in touch devices?

Author: Yes you're right, I will make the distinction

Author: Yes you're right, I will make the distinction

| Mouse    | Click      | Open    | Click the menu item that contains a submenu                            |
| Mouse    | Hover      | Open    | Hover over the menu item that contains a submenu with delay            |
| Keyboard | Enter      | Open    | Focus on triggering menu item                                          |
| Keyboard | Space      | Open    | Focus on triggering menu item                                          |
| Keyboard | ArrowRight | Open    | Focus on triggering menu item                                          |
| Mouse    | Click      | Dismiss | Click on an item in the submenu                                        |
| Mouse    | Click      | Dismiss | Click on a UI element that is not the submenu                          |
| Mouse    | MouseLeave | Dismiss | Mouse leaves the submenu or its triggering menu item after delay       |
| Keyboard | ArrowLeft  | Dismiss | Closes the submenu and focuses on the triggering menu item             |
| Keyboard | Esc        | Dismiss | Closes the submenu and focuses on the triggering menu item             |
| Keyboard | Tab        | Dismiss | Closes the menu and all submenus, focus moves to next element in order |

> 

Reviewer: also see https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html# for extensive description of left/right arrow keys for submenu interaction

Author: That is exactly where most of this comes from, I will add a reference link to the section

Author: Added a references section to this chapter

v0 suffers from a consistency issue that the control used in `Menu` and the menu variant of `ToolbarItem` are not actually the same component and have different behavior. However, semantically for the purposes of this spec, they representthe same control that will be implemented.

Note that the below code samples are not meant to be complete, but to highlight differences between the two libraries. Please refer to official docsites for actual API references.

### Positioning

`ContextualMenu` in v7 is a component that also exposes the API to control the positioning of the temporary popup surface that the Menu is rendered on. This aspect of the v7 component should be compared with the `Popup` component in v0 since the v0 `Menu` is created as a standalone component with no positioning properties.

v0 uses the [OSS Popper.js library](https://popper.js.org/) while v7 uses a component based implementation `CalloutContent`. Below we provide the results of testing common positioning boundary/edge cases between the two.

> // TODO compare boudary/edge cases in positioning

Reviewer 5: Also consider dual-screen support (#12785). Double check whether it is currently a requirement.

When an options is associated with a selection state. The `Menu`, either root or nested, should control its dismiss behavior accordingly based on configuration.

### Sections

A `Menu` can be partitioned into sections using visible dividers in its list of options. Each section can contain a heading title that announces or briefly describes the options in the particular section

### Secondary text

An option of a `Menu` component should be able to declare additional secondary text that can provide additional context describing the option or its usage.

For example a secondary text can be a label that shows a keyboard shortcut that will perform an equivalent action of the option of the `Menu` component.

### Split option with nesting

An option of a `Menu` component can trigger a nested `Menu` component and also perform its default action by splitting the option into two interactable areas that handle each action separately.

### Disabled option(s)

All options in a `Menu` component can be disabled and should provide a visible indication for this purpose. User interaction should be defined for disabled options

> ### Scrollable

Reviewer 5: Do we need to support application-provided custom JS scrollbars?

Author: I'm pretty sure we do, we need to support hover actions for a scrollbar which native browsers generally do really badly. This is also smth that @assuncaocharles has looked into already for v0, would love to get his experiences

### MenuSection

Creates a section inside a `MenuList`, setting up header layout and dividers between `MenuItems`.

The MenuSection is also a useful component to declare different selection groups (checkbox/radio) in a `MenuList`.

### MenuDivider

Creates a divider element in the `MenuList` with correct HTML and aria semantics for divider. Intention is to use this internally within `MenuList` and `MenuSection` and avoid explicit uses of dividers to ensure correct HTML and aria semantics.

### MenuItem

As the name infers

### SubMenu

Creates a `Menu` component with `MenuItem` trigger and handles the positioning of the nested menu.

## Sample code

The below samples do not represent the definitive props of the final implemented component, but represent the ideal final implementations. Can be subject to change during the implementation phase.

### Default Menu

```typescript
const trigger = <button> Open menu </button>

const menu = (
  <Menu trigger={trigger}>
    <MenuItem>Option 1</MenuItem>
    <MenuItem>Option 2</MenuItem>
    <MenuItem>Option 3</MenuItem>
  <Menu>
)
```

```html
<!-- expected DOM output  -->
<!-- TODO positioning -->
<div role="menu">

>   <div role="menuitem" tabindex="0">Option 1</div>

Reviewer 5: Have you verified a11y is ok with just `div`s?

Author: Yes thanks to @kolaps33

  <!-- TODO submenu -->
</div>
```

### Standlone

```typescript
const [open] = React.useState(false);

const menu = (
  <CustomSurface open={open}>
    <MenuList>
      <MenuItem>Option 1</MenuItem>
      <MenuItem>Option 2</MenuItem>
      <MenuItem>Option 3</MenuItem>
    <MenuList>
  <CustomSurface>
)
```

```html
<!-- expected DOM output  -->
<div role="menu">
  <div role="menuitem" tabindex="0">Option 1</div>
  <div role="menuitem" tabindex="-1">Option 2</div>
  <div role="menuitem" tabindex="-1">Option 3</div>
</div>
```

### Selection

```typescript
const trigger = <button> Open menu </button>
const [selectedItems, setSelectedItems] = React.useState([]);

// basic checkbox example
const menuCheckbox = (
  <Menu
    selectedItems={selectedItems}
    onSelectionChange={setSeelctedItems}
    trigger={trigger}
  >
    <MenuItem index={1}>Option 1</MenuItem>
    <MenuItem index={2}>Option 2</MenuItem>
    <MenuItem index={3}>Option 3</MenuItem>
  <Menu>
)

// leverage MenuSection for different selection groups
const menuSelectableSections = (
  <Menu
    selectedItems={selectedItems}
    onSelectionChange={setSeelctedItems}
    trigger={trigger}
  >

>     <MenuSection kind="checkbox">

Reviewer 5: What is this `MenuSection` good for?

Author: `MenuSection` would be responsible for adding dividers and section headers in a correct way in terms of html semantics and narration. It would also help manage create selectable groups of menu checkboxes and radios which require some extra DOM elements for a consisent narration experience

Reviewer 5: I see the additional wrapping element for radiogroup, I just wasn't sure about the checkbox section.

Author: Ah that's a mistake on my part will fix, my intention is that any selectable group in a `MenuSection` needs to be wrapped to provide a consistent narration experience. 

This also means that if you have a menu with only checkboxes it actually makes sense **not** to use `MenuSection`

Author: Added the missing group

      <MenuItem index={1}>Option 1</MenuItem>
      <MenuItem index={2}>Option 2</MenuItem>
      <MenuItem index={3}>Option 3</MenuItem>
    </MenuSection>
    <MenuSection kind="radio">
      <MenuItem index={4}>Option 1</MenuItem>
      <MenuItem index={5}>Option 2</MenuItem>
      <MenuItem index={6}>Option 3</MenuItem>
    </MenuSection>
  <Menu>
)
```

```html
<!-- expected DOM output for basic checkbox  -->
<div role="menu">

>   <div role="menuitemcheckbox" tabindex="0" aria-checked="true">Option 1</div>

Reviewer 5: When are the `role` and `aria-checked` added?

Author: This should be handled by each `MenuItem` from proposed context that would be used in the `Menu`

I definitely do need to add more detail about what context should do that's a good point

# Menu

## Background

### Definition

This spec defines the default function of a `Menu` as an interactive component that displays a list of options that can be represented by a range possible states. Possible variants are defined in [the relevant section](#variants)

The `Menu` should be displayed on a temporary surface that interrupts the normal flow of content. The temporary surface should be triggered by an external user action such as (but not limited to) a click on a button or other UI control.

The interactions that result in the dismiss/removal of the `Menu` component should be configurable.

## Prior art

As a part of the spec definitions in Fluent UI, a research effort has been made through [Open UI](https://open-ui.org/). The current research proposal is available as an open source contribution undergoing review ([research proposal](https://github.com/WICG/open-ui/pull/249))

## Comparison of `@fluentui/react` and `@fluentui/react-northstar`

- All mentions of v7 or v8 == `@fluentui/react` ([docsite](https://developer.microsoft.com/en-us/fluentui#/))
- All mentions of v0 == `@fluentui/react-northstar` ([docsite](https://fluentsite.z22.web.core.windows.net/))

> The most relevant comparison that can be achieved between the two libraries is between `ContextualMenu` in v7 and a combination of `Menu`, `Popup` and `ToolbarItem` in v0.

Reviewer 5: I would like to have more detailed mapping from converged to v7/8 and v0.
Once the converged `Menu` is implemented, what exactly will it replace in existing libraries?
According to design spec, for v0 it will not replace `Menu` but rather a `MenuButton` and perhaps some other scenarios?

Author: The spec just grows larger... :-) I will try to clarify this more

just FYI: I propose a `MenuList` component that can be adequately used to replace v0 `Menu` where as a `MenuButton` should be implemented using `Menu` pretty easily

in v7 the contender here is [ContextualMenu](https://developer.microsoft.com/en-us/fluentui#/controls/web/contextualmenu)

    <MenuItem>Option 1</MenuItem>
    <MenuSection title="Section title">
      <MenuItem>Section Option 1</MenuItem>
      <MenuItem>Section Option 2</MenuItem>
      <MenuItem>Section Option 3</MenuItem>
    <MenuSection />
  <Menu>
)
```

```html
<!-- expected DOM output  -->
<!-- TODO positioning -->
<button aria-haspopup="true" aria-expanded="true" id="trigger">Open menu</button>
<div role="menu" aria-labelledby="trigger">
  <div role="menuitem" tabindex="0">Option 1</div>
  <div role="separator"></div>
  <div role="presentation" aria-hidden="true" id="sectionid">Section title</div>
  <div role="group" aria-labelledby="sectionid">
    <div role="menuitem" tabindex="-1">Section Option 1</div>
    <div role="menuitem" tabindex="-1">Section Option 2</div>
    <div role="menuitem" tabindex="-1">Section Option 3</div>
  </div>
  <div role="separator"></div>
</div>
```

### Submenus

```typescript
const trigger = <button> Open menu </button>

const menu = (
  <Menu trigger={trigger}>
    <MenuItem>Option 1</MenuItem>
    <SubMenu content="Open submenu">
      <MenuItem>Option 1</MenuItem>
      <MenuItem>Option 2</MenuItem>
      <MenuItem>Option 3</MenuItem>

>     <SubMenu>

Reviewer 4: It'd be great to have the API for SubMenu be parallel to the API for MenuItem and Menu (with regards to content and triggering), rather than introducing the `content` slot for SubMenu only, and not for MenuItem.

If we went with the current `trigger` proposal, it would look something like:
```tsx
<Menu trigger={<Button>Open menu</Button>}>
  <MenuItem>Option 1</MenuItem>
  <SubMenu trigger={<MenuItem>Open submenu</MenuItem>}>
    <MenuItem>Option 1</MenuItem>
    <MenuItem>Option 2</MenuItem>
    <MenuItem>Option 3</MenuItem>
  </SubMenu>
</Menu>
```

Another option might be to have subMenu be a slot on MenuItem, and the content of the MenuItem would still be its child:
```tsx
<Menu trigger={<Button>Open menu</Button>}>
  <MenuItem>Option 1</MenuItem>
  <MenuItem subMenu={
      <SubMenu>
        <MenuItem>Option 1</MenuItem>
        <MenuItem>Option 2</MenuItem>
        <MenuItem>Option 3</MenuItem>
      </SubMenu>}>
    Open submenu
  </SubMenu>
</Menu>
```

Author: @behowell I recommend you come to the trigger deepdive !

I would avoid putting `MenuItems` in a slot. I would prefer no slots tbh. It's a 'sacrifice' I made in this API proposal to have some way of associating an extraneous element to the declaration of the actual menu as children. I would like a API structure to reflect DOM structure as much as possible.]

Several ppl have approached me to help me improve and they will be attending the deepdive. You should definitely come and weigh in too :-)

When an options is associated with a selection state. The `Menu`, either root or nested, should control its dismiss behavior accordingly based on configuration.

### Sections

A `Menu` can be partitioned into sections using visible dividers in its list of options. Each section can contain a heading title that announces or briefly describes the options in the particular section

### Secondary label

An option of a `Menu` component should be able to declare additional secondary label that can provide additional context describing the option or its usage.

For example a secondary label can be a label that shows a keyboard shortcut that will perform an equivalent action of the option of the `Menu` component.

### Split option with nesting

An option of a `Menu` component can trigger a nested `Menu` component and also perform its default action by splitting the option into two interactable areas that handle each action separately.

### Disabled option(s)

All options in a `Menu` component can be disabled and should provide a visible indication for this purpose. User interaction should be defined for disabled options

> ### Scrollable

Reviewer 6: Do we need to provide any kind of UI Virtualization hooks for scrolling long lists of menu items? One example, font picking scenarios where a user could have 300+ fonts on the system.

Author: That's a very good question, I'd love to schedule some time with @YuanboXue-Amber to talk about her recent efforts in virtualizing the v0 Tree with a hooks refactor and and try to leverage her experiences in how this proposed API could do in a similar situation

Author: Also from what I see neither v0 nor v7 currently support the concept of a virtualized menu, it's rather list that should be virtualized

Author: After and audit of some virtualization libraries and a talk with @YuanboXue-Amber who did virtualization recently for the v0 `Tree`, it seems a children based API will play nicely. Since this is also not supported in either v7 or v0, let's open that can of worms once we get there
